ops/main.py

```diff
@@ -356,9 +325,255 @@ def _should_use_controller_storage(db_path: Path, meta: CharmMeta) -> bool:
+            )
+        meta = yaml.safe_load(metadata_path.open())
+
+        actions = config = None
+
+        config_path = charm_root / "config.yaml"
+        if config_path.exists():
+            config = yaml.safe_load(config_path.open())
+
+        actions_path = charm_root / "actions.yaml"
+        if actions_path.exists():
+            actions = yaml.safe_load(actions_path.open())
+
+        return _CharmSpec(
+            charm_type=charm_type,
+            charm_root=charm_root,
+            meta=meta,
+            actions=actions,
+            config=config
+        )
+
+
+class Ops:
+    """Initializes the Framework and manages the lifecycle of a charm.
+
+    Running Ops consists of three main steps:
+    - setup: initialize from envvars and argv's the objects that encapsulate:
+      - the Framework (hook tool wrappers)
+      - the storage backend
+      - the event that Juju is emitting on us
+      - the charm instance (user-facing)
+    - emit: core user-facing lifecycle step. Consists of:
+      - reemit any deferred events found in the storage
+      - emit the Juju event on the charm
+      - emit any other events the charm emitted on itself during this cycle
+    - commit: responsible for:
+      - store any events deferred throughout this execution
+      - graceful teardown of the storage
+      - emission of the ``collect-status`` events
+    """
+    CHARM_STATE_FILE = '.unit-state.db'
+
+    def __init__(
+            self,
+            charm_spec: "_CharmSpec",
+            use_juju_for_storage: Optional[bool] = None
+    ):
+        self.charm_spec = charm_spec
+        self._charm_meta = CharmMeta(charm_spec.meta, charm_spec.actions)
+        self._use_juju_for_storage = use_juju_for_storage
+
+        # set by setup()
+        self.dispatcher: Optional[_Dispatcher] = None
+        self.framework: Optional["ops.framework.Framework"] = None
+        self.charm: Optional["ops.charm.CharmBase"] = None
+
+        self._has_setup = False
+        self._has_emitted = False
+        self._has_committed = False
+
+    def _setup_charm(self, framework: "ops.framework.Framework", dispatcher: _Dispatcher):
+        charm_type = self.charm_spec.charm_type
+        sig = inspect.signature(charm_type)
+        sig.bind(framework)  # signature check
+
+        charm = charm_type(framework)
+        dispatcher.ensure_event_links(charm)
+        return charm
+
+    @staticmethod
+    def _setup_model_backend():
+        # this is a standalone method to facilitate overriding by subclassing in testing
+        return ops.model._ModelBackend()
+
+    @staticmethod
+    def _setup_root_logging(backend: ops.model._ModelBackend):
+        debug = "JUJU_DEBUG" in os.environ
+        setup_root_logging(backend, debug=debug)
+
+        # our hello world
+        logger.debug(
+            "ops %s up and running.",
+            ops.__version__,
+        )  # type:ignore
```

Comment:
This easily fits on one line in the existing code, and one additional 4-space indent shouldn't change that. Did you wrap this, or was it a tool?